### PR TITLE
CLDSRV-375 Exclude already transitioned keys from lifecycle listings

### PR DESCRIPTION
**Pull request by nicolas2bert**
A new listing parameter called "excluded-data-store-name" is added to filter out keys that have already been transitioned. 

Note that orphan delete markers do not go through a transition workflow, so no filtering based on datastore name is applied to list orphan delete markers.

Note once https://github.com/scality/Arsenal/pull/2111 is merged, update the Arsenal package dependency.

### CODE REVIEW

File: lib/api/backbeat/listLifecycleOrphanDeleteMarkers.js
Comment (jonathan-gramain) [inline lib/api/backbeat/listLifecycleOrphanDeleteMarkers.js:31]: I believe we need to pass `locationConstraints` in the `listParams` here as well.

File: tests/functional/backbeat/excludedDataStoreName.js
Comment (jonathan-gramain) [inline tests/functional/backbeat/excludedDataStoreName.js:122]:
```suggestion
            assert.deepStrictEqual(data.Contents, [
                { Key: 'key0', DataStoreName: location1 },
                { Key: 'key2', DataStoreName: location1 }
            ]);
```

**Review by jonathan-gramain — CHANGES_REQUESTED** [review]

File: package.json
Comment (nicolas2bert) [inline package.json]: TODO: need to be updated before merging.

File: lib/api/backbeat/listLifecycleOrphanDeleteMarkers.js
Comment (nicolas2bert) [reply lib/api/backbeat/listLifecycleOrphanDeleteMarkers.js:31]: Good catch. I am wondering how the functional "listing orphan delete markers" tests did not catch this. We might want to give it a quick look. Thanks.

Comment (nicolas2bert) [reply lib/api/backbeat/listLifecycleOrphanDeleteMarkers.js:31]: I don't think so. What we pass in the `listParams`  is the `excludedDataStoreName` property that will be used to exclude keys that have already been transitioned. 
It is not necessary since orphan delete markers are not transitioned.

File: lib/api/backbeat/listLifecycleCurrents.js
Comment (jonathan-gramain) [inline lib/api/backbeat/listLifecycleCurrents.js:53]: This check seems to be duplicated in each listing backend that uses it, why not instead validating in `listLifecycle`, before calling the backend function? This would also avoid having to pass the `locationConstraints` object around.

File: lib/api/backbeat/listLifecycleOrphanDeleteMarkers.js
Comment (jonathan-gramain) [reply lib/api/backbeat/listLifecycleOrphanDeleteMarkers.js:31]: OK, makes sense, it would be useful to describe in the JSDoc of each listing backend function which params are supported and what they are for, to avoid surprises later if we add new params etc., otherwise one may assume that all backend functions support the same set of params, which is not the case here.

**Review by jonathan-gramain — APPROVED** [review]
Left some suggestions to improve a few things, not blocking the PR to be merged if in a time crush but ideally would be addressed before merging.

File: lib/api/backbeat/listLifecycleCurrents.js
Comment (nicolas2bert) [reply lib/api/backbeat/listLifecycleCurrents.js:53]: List orphans delete markers don't use it.

Comment (nicolas2bert) [reply lib/api/backbeat/listLifecycleCurrents.js:53]: The check is utilized by both the 'listLifecycleCurrents' and 'listLifecycleNonCurrents' API calls, but not by the 'listLifecycleDeleteMarkers' API call. To clarify that the check applies only to the former two calls, I moved it to the 'API level'.

Comment (jonathan-gramain) [reply lib/api/backbeat/listLifecycleCurrents.js:53]: Well, still not very clear to me ;)

I think for separation of concerns, it's better if the check on the locations happens from the API caller. It's ok if it is also checked for the "listLifecycleDeleteMarkers" case. Then, we pass the excluded (and validated) location to the backend listing, and it's up to the listing to decide whether to implement this exclusion mechanism or not (and document appropriately in JSDoc).

I believe it works as is but I find it a bit confusing and less easily maintainable.

**Review by alexanderchan-scality — APPROVED** [review]

File: lib/api/backbeat/listLifecycleCurrents.js
Comment (nicolas2bert) [reply lib/api/backbeat/listLifecycleCurrents.js:53]: For "listLifecycleDeleteMarkers" case, I think we should check that `params['excluded-data-store-name']` is `undefined` instead and we should return a different error than "getLocationConstraintErrorMessage" if `params['excluded-data-store-name']` is defined.

I have not implemented this check today since it is an internal API but I will create a ticket for it.

**Pull request merged**
